source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -12,70 +12,154 @@ namespace Extensions {
 namespace HttpFilters {
 namespace HeaderMutation {
 
-void Mutations::mutateRequestHeaders(Http::HeaderMap& headers,
-                                     const Formatter::HttpFormatterContext& ctx,
+void QueryParameterMutationAppend::mutateQueryParameter(
+    Http::Utility::QueryParamsMulti& params, const Formatter::HttpFormatterContext& context,
+    const StreamInfo::StreamInfo& stream_info) const {
+
+  const std::string value = formatter_->formatWithContext(context, stream_info);
+
+  switch (action_) {
+    PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
+  case ParameterAppendProto::APPEND_IF_EXISTS_OR_ADD:
+    params.add(key_, value);
+    return;
+  case ParameterAppendProto::ADD_IF_ABSENT: {
+    auto iter = params.data().find(key_);
+    if (iter == params.data().end()) {
+      params.add(key_, value);
+    }
+    break;
+  }
+  case ParameterAppendProto::OVERWRITE_IF_EXISTS: {
+    auto iter = params.data().find(key_);
+    if (iter != params.data().end()) {
+      params.overwrite(key_, value);
+    }
+    break;
+  }
+  case ParameterAppendProto::OVERWRITE_IF_EXISTS_OR_ADD:
+    params.overwrite(key_, value);
+    break;
+  }
+}
+
+Mutations::Mutations(const MutationsProto& config, absl::Status& creation_status) {
+  auto request_mutations_or_error = HeaderMutations::create(config.request_mutations());
+  SET_AND_RETURN_IF_NOT_OK(request_mutations_or_error.status(), creation_status);
+  request_mutations_ = std::move(request_mutations_or_error.value());
+
+  auto response_mutations_or_error = HeaderMutations::create(config.response_mutations());
+  SET_AND_RETURN_IF_NOT_OK(response_mutations_or_error.status(), creation_status);
+  response_mutations_ = std::move(response_mutations_or_error.value());
+
+  for (const auto& mutation : config.query_parameter_mutations()) {
+    if (mutation.has_append()) {
```

Comment:
Please add a comment here about the precedence of `append` over `remove` as it is not clear from the API. It would be better to clarify this in the API so all uses of this field will behave the same, but I'm not sure where this type is used elsewhere. (Better yet, if the type is used somewhere else consider consolidating the config handling code to a single function)

Reply:
Update the comment and the code to use strict oneof semantics first.

> (Better yet, if the type is used somewhere else consider consolidating the config handling code to a single function)

I didn't think of a good position for the code for now. let's defer this to future.